update sa receiver readme (#2513)
signalfx/splunk-otel-collector: pkg/receiver/smartagent/README.md

# SignalFx Smart Agent Receiver

> **Note:** The SignalFx Smart Agent receiver is only fully supported on x86_64/amd64 platforms.

> The Smart Agent Receiver allows you to use [SignalFx Smart Agent monitors](https://github.com/signalfx/signalfx-agent#monitors)

Reviewer: ```suggestion
The Smart Agent Receiver lets you use [SignalFx Smart Agent monitors](https://github.com/signalfx/signalfx-agent#monitors)
```

in the [Splunk Distribution of OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector). Many
monitors also require a [Smart Agent release bundle](https://github.com/signalfx/signalfx-agent/releases/latest),

> which is installed by the Splunk Distribution of OpenTelemetry Collector installer on supported x86_64/amd64 platforms.

Reviewer: ```suggestion
which is installed by the Splunk Distribution of OpenTelemetry Collector on supported x86_64/amd64 platforms.
```

> While this document outlines some of the capabilities of the Smart Agent receiver, please see the

Reviewer: ```suggestion
See the
```

[migration guide](../../../docs/signalfx-smart-agent-migration.md)
for more information about migrating from the Smart Agent to the Splunk Distribution of the OpenTelemetry Collector.

> **Beta: All Smart Agent monitors are supported at this time. Configuration and behavior may change without notice.**

Reviewer: ```suggestion
**Beta: All Smart Agent monitors are supported by Splunk. Configuration and behavior may change without notice.**
```

## Configuration

For each Smart Agent 
[monitor](https://github.com/signalfx/signalfx-agent/blob/main/docs/monitor-config.md)

> you want to put into the Collector, add a `smartagent` receiver configuration block. Once configured in the Collector, each

Reviewer: "add to" instead of "put into"?

Author: Better. Fixed throughout.

> `smartagent` receiver will act as a drop-in replacement for its corresponding Smart Agent monitor.

Reviewer: ```suggestion
`smartagent` receiver acts as a replacement for its corresponding Smart Agent monitor.
```

Author: fixed

> 1. Put any Smart Agent or collectd configuration into the global

Reviewer: This is a rather long list of options. Consider splitting into subsections and add an example for each (or link to one)

Author: Thanks, mind if we hand this off to your team?

[Smart Agent Extension section](../../extension/smartagentextension/README.md)
of your Collector configuration.

> 1. Instead of using Smart Agent's `discoveryRule`s, use the Collector's

Reviewer: ```suggestion
1. Instead of using `discoveryRule`, use the Collector's
```

[Receiver Creator](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/receivercreator/README.md)
and [Observer extensions](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/README.md).

> 1. If using a [SignalFx Forwarder](https://github.com/signalfx/signalfx-agent/blob/main/docs/monitors/signalfx-forwarder.md)

Reviewer: ```suggestion
1. If you're using a [SignalFx Forwarder](https://github.com/signalfx/signalfx-agent/blob/main/docs/monitors/signalfx-forwarder.md)
```

monitor, put it into both a `traces` and a `metrics` pipeline, and use a
[Sapm exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/sapmexporter/README.md)
and a 
[SignalFx exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/signalfxexporter/README.md),
as each pipeline's exporter, respectively.
1. To replace or modify metrics, use
[Collector processors](https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/README.md).

> 1. If you have a monitor that sends [events](https://dev.splunk.com/observability/docs/datamodel/custom_events) (e.g. `kubernetes-events`,
> `nagios`, `processlist`, and some `telegraf` monitors like `telegraf/exec`), put it in a `logs` pipeline that uses a
> [SignalFx exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/signalfxexporter/README.md).
> It's recommended, and in the case of the Processlist monitor required, to put into the same pipeline a
> [Resource Detection processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/README.md),
> which will add host information and other useful dimensions to the events. An example is provided below.

Reviewer: ```suggestion
1. If you have a monitor that sends [events](https://dev.splunk.com/observability/docs/datamodel/custom_events), for example `kubernetes-events`,
`nagios`, `processlist`, and some `telegraf` monitors like `telegraf/exec`, put it in a `logs` pipeline that uses a
[SignalFx exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/signalfxexporter/README.md).
If the Processlist monitor is required, add to the same pipeline a
[Resource Detection processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/README.md),
which adds host information and other useful dimensions to the events. See the example.
```

I suggest linking to the example. Positional language ("below") isn't very useful in this case.

Author: Mind if we hand this off to your team?

> 1. If you have a monitor that updates [dimension properties or tags](https://dev.splunk.com/observability/docs/datamodel/metrics_metadata)
> (e.g. `ecs-metadata`, `heroku-metadata`, `kubernetes-cluster`, `openshift-cluster`, `postgresql`, or `sql`), put the name of
> your SignalFx exporter in its `dimensionClients` field in the Collector's SignalFx receiver configuration block.
> If you do not specify any exporters in this array field, the receiver will attempt to use the Collector pipeline to which it's connected. If
> the next element of the pipeline isn't compatible with updating dimensions, and if you configured a single SignalFx exporter,
> the reciever will use that SignalFx exporter. If you don't require dimension updates, you can specify the empty array `[]` to disable it.

Reviewer: ```suggestion
1. If you have a monitor that updates [dimension properties or tags](https://dev.splunk.com/observability/docs/datamodel/metrics_metadata), for example `ecs-metadata`, `heroku-metadata`, `kubernetes-cluster`, `openshift-cluster`, `postgresql`, or `sql`, put the name of
your SignalFx exporter in its `dimensionClients` field in the Collector's SignalFx receiver configuration block.
If you don't specify any exporters in this array field, the receiver attempts to use the Collector pipeline to which it's connected. If
the next element of the pipeline isn't compatible with updating dimensions, and if you configured a single SignalFx exporter,
the receiver uses that SignalFx exporter. If you don't require dimension updates, you can specify the empty array `[]` to disable it.
```

> **Note:** The SignalFx Smart Agent receiver is only supported on x86_64/amd64 platforms.

The Smart Agent Receiver allows you to utilize existing [SignalFx Smart Agent monitors](https://github.com/signalfx/signalfx-agent#monitors)
as OpenTelemetry Collector metric receivers.  It assumes that you have a properly configured environment with a
functional [Smart Agent release bundle](https://github.com/signalfx/signalfx-agent/releases/latest) on your system,

> which is already provided for x86_64/amd64 Splunk Distribution of OpenTelemetry Collector

Author: But the note above says `The SignalFx Smart Agent receiver is only supported on x86_64/amd64 platforms.` Should we mention that again here?

> which is already provided for x86_64/amd64 Splunk Distribution of OpenTelemetry Collector

Reviewer 2: sorry, specifically the "installation paths" since the distribution executable doesn't provide them alone.

> which is already provided for x86_64/amd64 Splunk Distribution of OpenTelemetry Collector

Author: Sorry, I'm a bit confused. Are we trying to address people installing a collector executable on a non-x86/amd64 host, which already had a Smart Agent installed on it? Is this supported?

I'm also confused about what is meant by installation paths, and the linked document doesn't mention them. Can you elaborate?

Perhaps if you could rephrase the statement it would help me understand.

> which is already provided for x86_64/amd64 Splunk Distribution of OpenTelemetry Collector

Reviewer 2: yeah, it's ambiguous overall. I'm not sure how accurate "only supported on x86_64/amd64 platforms" is as a whole since the native go monitors can work for other arches but don't think that'd need to be addressed here.

The SA bundle is only deployed by the x86 installers (and made available to the docker images) and the link to the getting started guide is, I believe, the intro to packaging and installation methods that provide the bundle. ARM or other arches users may be surprised by the bundle not being available to them after installing the rpm, for example, and calling that out here helps provide some expectation setting.

> which is already provided for x86_64/amd64 Splunk Distribution of OpenTelemetry Collector

Author: I see, thanks. I've added some language that hopefully helps folks understand.

> which is already provided for x86_64/amd64 Splunk Distribution of OpenTelemetry Collector

Reviewer 2: thanks, this is clearer than the original.

While this document outlines some of the capabilities of the Smart Agent receiver, please see the
[migration guide](../../../docs/signalfx-smart-agent-migration.md)
for more information about migrating from the Smart Agent to the Splunk Distribution of the OpenTelemetry Collector.

**Beta: All Smart Agent monitors are supported at this time. Configuration and behavior may change without notice.**

## Configuration

For each Smart Agent 
[monitor](https://github.com/signalfx/signalfx-agent/blob/main/docs/monitor-config.md)
you want to put into the Collector, add a `smartagent` receiver configuration block. Once configured in the Collector, each
`smartagent` receiver will act as a drop-in replacement for its corresponding Smart Agent monitor.

> 1. Put any Smart Agent or collectd configuration into the

Reviewer 2: I think stressing "global" is helpful here since "any smart agent or collectd configuration" isn't accurate if users assume an implicit monitor inclusion.

Author: Sure, done.

[Smart Agent Extension section](../../extension/smartagentextension/README.md)
of your Collector configuration.
1. Instead of using Smart Agent's `discoveryRule`s, use the Collector's
[Receiver Creator](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/receivercreator/README.md)
and [Observer extensions](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/README.md).

> 1. Instead of using the [SignalFx Forwarder](https://github.com/signalfx/signalfx-agent/blob/main/docs/monitors/signalfx-forwarder.md)

Reviewer 2: The signalfx-forwarder monitor in both metrics and trace pipelines was the important bit and it isn't replaced by the   sapm exporter.

Author: Nice catch. Thanks.

in a `traces` pipeline, and make sure you have a
[SignalFx exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/signalfxexporter/README.md)
configured and in a `metrics` pipeline. (meta: add a sentence as to why)
1. To replace or modify metrics, use
[Collector processors](https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/README.md).
1. If you have a monitor that updates [dimension properties or tags](https://dev.splunk.com/observability/docs/datamodel/metrics_metadata)
(e.g. `ecs-metadata`, `heroku-metadata`, `kubernetes-cluster`, `openshift-cluster`, `postgresql`, or `sql`), put the name of
your SignalFx exporter in its `dimensionClients` field in the Collector's SignalFx receiver configuration block.
If you do not specify any exporters in this array field, the receiver will attempt to use the Collector pipeline to which it's connected. If
the next element of the pipeline isn't compatible with updating dimensions, and if you configured a single SignalFx exporter,
the reciever will use that SignalFx exporter. If you don't require dimension updates, you can specify the empty array `[]` to disable it.

> 1. If you have a monitor that sends [events](https://dev.splunk.com/observability/docs/datamodel/custom_events) (e.g. `kubernetes-events`,

Reviewer 2: mind moving this up one in importance since this is required and the dimension client logic has automatic defaults?

Author: Sure, done.

**Beta: All Smart Agent monitors are supported at this time. Configuration and behavior may change without notice.**

## Configuration

For each Smart Agent 
[monitor](https://github.com/signalfx/signalfx-agent/blob/main/docs/monitor-config.md)
you want to put into the Collector, add a `smartagent` receiver configuration block. Once configured in the Collector, each
`smartagent` receiver will act as a drop-in replacement for its corresponding Smart Agent monitor.

1. Put any Smart Agent or collectd configuration into the global
[Smart Agent Extension section](../../extension/smartagentextension/README.md)
of your Collector configuration.
1. Instead of using Smart Agent's `discoveryRule`s, use the Collector's
[Receiver Creator](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/receivercreator/README.md)
and [Observer extensions](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/README.md).
1. If using a [SignalFx Forwarder](https://github.com/signalfx/signalfx-agent/blob/main/docs/monitors/signalfx-forwarder.md)

> monitor, put it into both a `metrics` and `traces` pipeline, and use a

Reviewer 2: nit to respect exporter ordering

```suggestion
monitor, put it into both a `traces` and `metrics` pipeline, and use a
```

Author: Good catch. Fixed.
